### DISCUSSION TITLE
Road to 1.0

### DISCUSSION THREAD

**Discussion by stephanbuys** [stephanbuys/docker-compose-types#9] [Q&A]
As end-users of the the crate, what would you expect to see for the crate to be marked as stable, it's a pretty simple data structure crate, hence it should be relatively easy to get there.

Ideas that come to mind:
- Documentation
- "Completeness", do we support all the fields and functionality?
- Better error handling on parsing failures
- Newtypes and enums to validate "stringly" data fields (ie. limit String fields to specific values where applicable)
- Others?

**Comment by taorepoara** [comment]
A Github action pipeline would be nice to build the sources, run the tests a generate the releases. You can find one here: https://github.com/lenra-io/dofigen/blob/main/.github/workflows/build_ci.yml (we used semantic release for the versioning, but it could be removed).

I also think that using `indexmap` is a bad idea since you force the projects using your crate to add it as dependency. You should use the standard `HashMap` instead since the order has no impact on the Docker Compose definition (at least I've not seen any).

**Reply by stephanbuys** [reply]
Happy to consider a github action in a PR. Currently I don't have the need or bandwidth to wire that up, but if someone wants to build that I'd happily consider incorporating it.

**Reply by taorepoara** [reply]
I will see to create a PR but you will have to add a crates.io token in the repo secrets. We will talk about this in the PR.

**Comment by taorepoara** [comment]
I think that "Completeness" would be nice two. It could be done with [schemafy](https://github.com/Marwes/schemafy) and the [Docker Compose JSON Schema](https://github.com/compose-spec/compose-spec/blob/master/schema/compose-spec.json).

**Reply by taorepoara** [reply]
I've tried to make a PoC of this, but schemafy does not manage the Docker Compose JSON Schema file version. This JSON Schema file can help to handle "Completeness" but not automatically.

**Reply by stephanbuys** [reply]
One way to handle this could be through a simple "utility" script to just generate the code, the schema is not extremely complex.

**Comment by taorepoara** [comment]
I also think that the structures should implement the Default trait through the derive tag.

**Reply by stephanbuys** [reply]
So some of the fields don't have obvious defaults or might need a review of the schema (or making more fields optional), but I'm happy to consider a PR that implements defaults.

**Reply by taorepoara** [reply]
Yes, I think that many fields could be optional.

**Reply by taorepoara** [reply]
Hi,
I'm working on this issue, but adding the attribute `#[serde(skip_serializing_if = "Option::is_none")]` to every optional field is a bit long and make the code less readable. 
What do you think of adding this on structs instead ? https://docs.rs/serde_with/latest/serde_with/attr.skip_serializing_none.html

**Reply by stephanbuys** [reply]
Sure, I've used `serde_with` in other crates before. Honestly, because there won't be that much churn inside of the data structures I would also consider the value of having less dependencies, and being way more explicit. So it would be a trade-off. But that said, there are so many optional fields that maybe it is justified regardless.

**Reply by taorepoara** [reply]
Yeah that right. I make my PR without it

**Comment by stephanbuys** [comment]
With the last few merges I think the crate has really started to see a good level of spec adoption as well as real-world testing. Not much holding back a v1.0.0 release, would love to know the community's views.

**Reply by taorepoara** [reply]
I think that the project seems ready but not the repo.

A little more documentation could be nice in the code to fill at least the main documentation page: https://docs.rs/docker-compose-types/0.5.0/docker_compose_types/

Automating tests and checks in the CI to be prevent problems before merging. I can take care of that part if you wish by reusing what we've made for the [Dofigen project](https://github.com/lenra-io/dofigen/blob/main/.github/workflows/build_ci.yml) or [our CLI](https://github.com/lenra-io/lenra_cli/blob/beta/.github/workflows/build_ci.yml) (that uses this project :smiley: )

**Comment by stephanbuys** [comment]
Hi Thomas, those are good points. So as far as a polished Open Source project is concerned I very much agree, as far as what is strictly meant by 1.0.0 (ie. stable, ready for production) I think these are worthwhile but not blocking issues.

I'm really glad this modest library has been useful to you :) 

If you have the bandwidth and the desire I would of course really welcome any contributions that bring us closer to the ideal. I'm currently not using the crate myself, so I'm just playing the role of custodian and release manager and not spending a lot of time on polish.
